dashboard/app/models/user.rb

```
@@ -620,8 +620,16 @@ def user_level_locked?(script_level, level)
   def next_unpassed_progression_level(script)
     user_levels_by_level = user_levels_by_level(script)
 
-    # Find the user level that we've most recently had progress on
-    user_level = user_levels_by_level.values.flatten.sort_by!(&:updated_at).last
+    # some of our user_levels may be for levels within level_groups, or for levels
+    # that are no longer in this script. we want to ignore those, and only look
+    # user_levels that have matching script_levels
+    levels = script.script_levels.map(&:level_ids).flatten
+    user_levels_by_level = user_levels.
+      where(script_id: script.id, level: levels).
+      index_by(&:level_id)
+
+    # Find the user_level that we've most recently had progress on
+    user_level = user_levels_by_level.values.max_by(&:updated_at)
```

New approach. Instead of calling script_levels on each level (which was a db hit each time), we start out by only getting user_levels for which we have a script_level. The call to `script.script_levels` should be cached, as is the call to level_ids on each script_level.

Now, on line 636 we're guaranteed to have last_script_level by non-nil, and thus our return will be non-nil

---

It's also worth noting that this method is likely failing silently in many cases in production right now. Most of the time, when we fail to find a `last_script_level`, we're okay because there is at least one unpassed progression level elsewhere in the script. In the case where that is not true, this method returns nil, and then we get HB errors.

---

I'll defer to Asher on the performance implications of adding all the level ids to the db query, but my gut feeling is that adapting your previous approach to use the cache would be more efficient (and doesn't seem *that* hacky to me).

---

The previous approaches cache problem was not where we thought. `user_level.level` and `Level.find_cache(user_level.level_id)` both end up hitting the cache. However, with both of those, when you end up doing `level.script_levels` we do not have a cache to hit.

This new approach should be much better at avoiding non-cached db hits.

---

My gut agrees with @balderdash, I think the other approach may be slightly more efficient. That said, I'm fairly convinced this method is efficient enough, so I don't think further changes are necessary. And I really like how it reads.

If you remind me to monitor the DB after this goes in, I'll look at whether this is worth optimizing. Regardless, I'm convinced it won't hurt production noticeably (as the non-cached `script_level` previously would have).